Firestore/Source/Local/FSTPersistence.h

```diff
@@ -99,6 +100,39 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)commitGroup:(FSTWriteGroup *)group;
 
+@property (nonatomic, readonly, assign) const FSTTransactionRunner& run;
+
+@end
+
+@protocol FSTTransactional
+
+- (void)startTransaction;
+
+- (void)commitTransaction;
+
 @end
 
+struct FSTTransactionRunner {
+
+// Intentionally disable nullability checking for this function. We cannot properly annotate
+// the function because this function can handle both pointer and non-pointer types. It is an error
+// to annotate non-pointer types with a nullability annotation.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnullability-completeness"
+  template <typename F>
+  auto operator() (F block) const -> decltype(block()) {
+    using ReturnT = decltype(block());
+    if (_db) {
+      [_db startTransaction];
+    }
+    ReturnT result = block();
+    if (_db) {
+      [_db commitTransaction];
+    }
+    return result;
```

Comment:
Just an observation: I think this line won't compile if `ReturnT` happened to be an rvalue-reference. I don't think it's really a problem, though, because it's pretty rare for a function to return `Foo&&`.

Reply:
I agree, and I'm inclined to not support that use-case until we have a concrete need for it.